File: python/leetcode/22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        results = []
        self.search(n, n, [], results)
        return results
    
    def search(self, left, right, path, results):
        if left == 0 and right == 0:
            results.append(''.join(path))
            return
        
        if left > 0:
            path.append('(')
            self.search(left - 1, right, path, results)
            path.pop()
            
        if left < right:
            path.append(')')
            self.search(left, right - 1, path, results)
            path.pop()

File: cpp/leetcode/22. Generate Parentheses.cpp
#include "vector"
using std::vector;
#include "string"
using std::string;

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> results;
        generate(n, n, "", results);
        return results;
    }
private:
    void generate(int left, int right, string str, vector<string>& results) {
        // base cases
        if (left == 0 && right == 0) {
            results.push_back(str);
            return;
        }    
        
        // recursive rules
        if (left > 0) {
            generate(left - 1, right, str + "(", results);
        }
        if (left < right) {
            generate(left, right - 1, str + ")", results);
        }
    }
};
